numpy/core/src/multiarray/compiled_base.c

```diff
@@ -825,29 +830,32 @@ static int sequence_to_arrays(PyObject *seq,
     }
 
     for (i = 0; i < count; ++i) {
-        PyObject *item = PySequence_GetItem(seq, i);
+        /* ensure dtype is based off dtype_guess */
+        PyArray_Descr *dtype = dtype_guess;
+        item = PySequence_GetItem(seq, i);
         if (item == NULL) {
-            while (--i >= 0) {
-                Py_DECREF(op[i]);
-                op[i] = NULL;
-            }
-            return -1;
+            goto fail;
         }
-
-        op[i] = (PyArrayObject *)PyArray_FROM_O(item);
+        Py_INCREF(dtype);
+        if (PyArray_DTypeFromObject(item, NPY_MAXDIMS, &dtype) < 0) {
+            goto fail;
```

Comment:
PyArray_DTypeFromObject does not fail. Perhaps we could improve the error message by detecting a non-integer dtype here. Currently the error is raised later with `TypeError: Iterator operand 0 dtype could not be cast from dtype('float64') to dtype('int64') according to the rule 'same_kind'`

Reply:
improved error message

Reply:
removed in a later commit